USPTO

baberabb opened the pull request:
This PR adds the code to process the USPTO dataset extracted from [Google Patents Public Dataset](https://cloud.google.com/blog/topics/public-datasets/google-patents-public-datasets-connecting-public-paid-and-private-patent-data) and uploaded to HF. The dataset covers all US patent applications until Oct 27, 2023 (including historical).

closes #9 

Edit: switched over to pandoc as that does the latex formatting pretty well, and also works well with the overall html.

blester125 commented on uspto/process_uspto.sh:
```diff
@@ -0,0 +1,44 @@
+cd mathml-to-latex || exit
+
+# Install Node.js dependencies
+npm install
```
I think we should segment this into two scripts, one that sets up the mathml-to-latex server (does the clone, install, etc) and one that runs data processing (start the server, run the code, kill the server, etc)

I'll do a more through review later today

baberabb commented on uspto/uspto-to-dolma.py:
```diff
@@ -0,0 +1,228 @@
+    Parameters:
+    - `data_dir` (str): The directory where the dataset is located. Default value is "./data/uspto/".
+    - `url` (str): The API endpoint URL for converting the dataset files. Default value is "http://localhost:3000/convert".
+    - `limit` (int): The maximum number of files to convert. Default value is 0, which means convert all files in the dataset.
```
yeah. good catch!

baberabb commented on uspto/uspto-to-dolma.py:
```diff
@@ -0,0 +1,228 @@
+        limit //= len(file_names)
+        logger.info(f"Processing {limit} entries each from {len(file_names)} files.")
+    args = [(x, url, limit) for x in file_names]
+    with multiprocessing.get_context("spawn").Pool() as pool:
```
So I'm mainly relying on `polars` here, which defaults to using all available cores. It usually works well enough but the API calls act as a blocker so using multiprocessing to process each file and let polars handle each individual file.

baberabb commented:
Thanks for the review @blester125! I think i have addressed all your points but let me know if I missed anything! Added the readme as well. The only thing left are the tables. They are mostly fine but the whitespace formatting is quite off. Will bring it up in the meeting!

blester125 commented on uspto/uspto-to-dolma.py:
````diff
@@ -0,0 +1,228 @@
+    ```
+    """
+
+    file_names = glob.glob(os.path.join(data_dir + "*.parquet"))
````
Nit: You shouldn't need the os.path.join here right? `data_dir` is typed as a string so it shouldn't do anything?

blester125 replied:
Thanks! Should be good to merge!

baberabb commented:
> Thanks! Feel free to merge when ready!

Whenever you're ready!
